# Bank Marketing Campaign
by : Jeffrey Junior Tedjasulaksana

# Import Library

In [3]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

# Import Data

Source Data :
https://www.kaggle.com/datasets/volodymyrgavrysh/bank-marketing-campaigns-dataset


In [4]:
# Import dataset
df = pd.read_excel('bank-additional-full.xlsx')
df

age          job  marital            education  default housing loan  \
0       56    housemaid  married             basic.4y       no      no   no   
1       57     services  married          high.school  unknown      no   no   
2       37     services  married          high.school       no     yes   no   
3       40       admin.  married             basic.6y       no      no   no   
4       56     services  married          high.school       no      no  yes   
...    ...          ...      ...                  ...      ...     ...  ...   
41183   73      retired  married  professional.course       no     yes   no   
41184   46  blue-collar  married  professional.course       no      no   no   
41185   56      retired  married    university.degree       no     yes   no   
41186   44   technician  married  professional.course       no      no   no   
41187   74      retired  married  professional.course       no     yes   no   

         contact month day_of_week  ...  campaign  pdays  previous  \
0      telephone   may         mon  ...         1    999         0   
1      telephone   may         mon  ...         1    999         0   
2      telephone   may         mon  ...         1    999         0   
3      telephone   may         mon  ...         1    999         0   
4      telephone   may         mon  ...         1    999         0   
...          ...   ...         ...  ...       ...    ...       ...   
41183   cellular   nov         fri  ...         1    999         0   
41184   cellular   nov         fri  ...         1    999         0   
41185   cellular   nov         fri  ...         2    999         0   
41186   cellular   nov         fri  ...         1    999         0   
41187   cellular   nov         fri  ...         3    999         1   

          poutcome emp.var.rate  cons.price.idx  cons.conf.idx  euribor3m  \
0      nonexistent          1.1         93994.0          -36.4     4857.0   
1      nonexistent          1.1         93994.0          -36.4     4857.0   
2      nonexistent          1.1         93994.0          -36.4     4857.0   
3      nonexistent          1.1         93994.0          -36.4     4857.0   
4      nonexistent          1.1         93994.0          -36.4     4857.0   
...            ...          ...             ...            ...        ...   
41183  nonexistent         -1.1         94767.0          -50.8     1028.0   
41184  nonexistent         -1.1         94767.0          -50.8     1028.0   
41185  nonexistent         -1.1         94767.0          -50.8     1028.0   
41186  nonexistent         -1.1         94767.0          -50.8     1028.0   
41187      failure         -1.1         94767.0          -50.8     1028.0   

       nr.employed    y  
0           5191.0   no  
1           5191.0   no  
2           5191.0   no  
3           5191.0   no  
4           5191.0   no  
...            ...  ...  
41183       4963.6  yes  
41184       4963.6   no  
41185       4963.6   no  
41186       4963.6  yes  
41187       4963.6   no  

[41188 rows x 21 columns]

# Data Understanding

Database Bank marketing campaigns : data yang menggambarkan hasil kampanye pemasaran bank Portugal. Kampanye yang dilakukan sebagian besar didasarkan pada panggilan telepon langsung, menawarkan klien bank untuk menempatkan deposito.

# Define Problem :

Menambah jumlah nasabah untuk melakukan deposito

# Define Goals :

Cara menambah jumlah nasabah untuk melakukan deposito

# Database Infromation :

- Age   	: Menyimpan informasi tentang umur klien
- Job   	: Menyimpan informasi tentang bidang pekerjaan klien
- Marital 	: Menyimpan informasi tentang status pernikahan klien
- Education 	: Menyimpan informasi tentang status pendidikan klien
- Default 	: Menyimpan informasi tentang latar belakang kredit macet klien
- Campaign	: Informasi terkait jumlah kontak yang dilakukan selama campaign ini
- Housing 	: Menyimpan informasi tentang cicilan rumah klien
- Loan 	: Menyimpan informasi tentang kepemilikan hutang klien terhadap pihak lain
- Contact 	: Menyimpan informasi tentang tipe menghubungi klien melalui telepon dan cellular
- Month 	: Menyimpan informasi tentang bulan menghubungi klien
- Day_of_Week : Menyimpan informasi tentang hari terakhir menghubungi klien
- Duration 	: Menyimpan informasi tentang durasi menghubungi klien melalui telepon dan cellular (dalam hitungan detik)
- Pdays : Menyimpan informasi tentang jumlah hari yang berlalu setelah klien terakhir dihubungi dari kampanye sebelumnya
- Previous : Menyimpan informasi tentang jumlah komunikasi dengan klien yang dilakukan yang dilakukan sebelum kampanye ini dan untuk klien ini
- Poutcome: Menyimpan informasi tentang hasil dari kampanye pemasaran sebelumnya
- Emp.Var.Rate : Menyimpan informasi tentang variasi tingkat lapangan kerja regional di suatu negara
- Cons.Price.Idx : Menyimpan informasi tentang tingkat inflasi di portugal
- Cons.Conf.Id : Menyimpan informasi tentang kepercayaan klien terhadap bank
- Euriborn3m : Menyimpan informasi tentang tingkat bunga rata-rata Europa
- Nr.employeed : Menyimpan informasi tentang jumlah karyawan bank
- Y : Menyimpan informasi tentang data klien yang jadi melakukan deposit di bank

# Data Wrangling

In [5]:
# Function age_range
def umur(x):
    if x <= 20 :
        return '< 21'
    elif x > 20 and x <=30 :
        return '21-30'
    elif x > 30 and x <=40 :
        return '31-40'
    elif x > 40 and x <=50 :
        return '41-50'
    elif x > 50 and x <=60 :
        return '51-60'
    elif x > 60 and x <=70 :
        return '61-70'
    elif x > 70 and x <=80 :
        return '71-80'
    elif x > 80 and x <=90 :
        return '81-90'
    else :
        return '>90'

In [6]:
# Insert Kolom age_year
df.insert(1, column='age_range',value=df['age'].apply(umur))

In [7]:
# Function edu
def edu(x):
    if x == 'basic.4y' :
        return 'basic'
    elif x == 'basic.6y' :
        return 'basic'
    elif x == 'basic.9y' :
        return 'basic'
    else :
        return x

In [8]:
# Insert Kolom edu
df.insert(5, column='edu',value=df['education'].apply(edu))

In [9]:
# Function campaign_range
def campaign_range(x):
    if x <= 10 :
        return '1-10'
    elif x > 10 and x <=20 :
        return '11-20'
    elif x > 20 and x <=30 :
        return '21-30'
    elif x > 30 and x <=40 :
        return '31-40'
    else :
        return '> 40'

In [10]:
# Insert Kolom campaign_range
df.insert(14, column='campaign_range',value=df['campaign'].apply(campaign_range))

In [11]:
# Function filter data 2009
def tahun(x):
    if x == 92843 :
        return '2009'
    elif x == 93074 :
        return '2009'
    elif x == 92893 :
        return '2009'
    elif x == 92963 :
        return '2009'
    elif x == 92469 :
        return '2009'
    elif x == 92201 :
        return '2009'
    elif x == 92379 :
        return '2009'
    elif x == 92431 :
        return '2009'
    elif x == 92649 :
        return '2009'
    elif x == 92731 :
        return '2009'
    else :
        return 'unknown'

In [12]:
# Insert Column Year
df.insert(11, column='year',value=df['cons.price.idx'].apply(tahun))

In [13]:
# Drop baris yang bukan tahun 2009
df=df.drop(df[df['year']!='2009'].index)
df

age age_range          job   marital          education  \
27690   26     21-30      student    single           basic.9y   
27691   38     31-40       admin.    single  university.degree   
27692   26     21-30      student    single           basic.9y   
27693   26     21-30      student    single           basic.9y   
27694   30     21-30   management   married  university.degree   
...    ...       ...          ...       ...                ...   
38953   81     81-90    housemaid  divorced           basic.4y   
38954   62     61-70  blue-collar   married           basic.4y   
38955   62     61-70  blue-collar   married           basic.4y   
38956   31     31-40  blue-collar   married           basic.6y   
38957   62     61-70  blue-collar   married           basic.4y   

                     edu default housing loan    contact  ... campaign_range  \
27690              basic      no     yes   no   cellular  ...           1-10   
27691  university.degree      no     yes   no   cellular  ...           1-10   
27692              basic      no      no   no   cellular  ...           1-10   
27693              basic      no     yes   no  telephone  ...           1-10   
27694  university.degree      no     yes   no  telephone  ...           1-10   
...                  ...     ...     ...  ...        ...  ...            ...   
38953              basic      no     yes   no   cellular  ...           1-10   
38954              basic      no      no   no   cellular  ...           1-10   
38955              basic      no     yes  yes   cellular  ...           1-10   
38956              basic      no     yes   no   cellular  ...           1-10   
38957              basic      no     yes  yes   cellular  ...           1-10   

      pdays previous     poutcome  emp.var.rate cons.price.idx  cons.conf.idx  \
27690   999        0  nonexistent          -1.8        92843.0          -50.0   
27691   999        0  nonexistent          -1.8        92843.0          -50.0   
27692   999        0  nonexistent          -1.8        92843.0          -50.0   
27693   999        0  nonexistent          -1.8        92843.0          -50.0   
27694   999        0  nonexistent          -1.8        92843.0          -50.0   
...     ...      ...          ...           ...            ...            ...   
38953   999        0  nonexistent          -3.4        92649.0          -30.1   
38954   999        1      failure          -3.4        92649.0          -30.1   
38955   999        0  nonexistent          -3.4        92649.0          -30.1   
38956   999        2      failure          -3.4        92649.0          -30.1   
38957   999        1      failure          -3.4        92649.0          -30.1   

       euribor3m nr.employed    y  
27690   1811.000      5099.1  yes  
27691   1811.000      5099.1  yes  
27692   1811.000      5099.1   no  
27693   1811.000      5099.1   no  
27694   1811.000      5099.1   no  
...          ...         ...  ...  
38953      0.719      5017.5   no  
38954      0.719      5017.5   no  
38955      0.719      5017.5   no  
38956      0.719      5017.5   no  
38957      0.719      5017.5   no  

[8810 rows x 25 columns]

In [14]:
# Drop kolom yang tidak terpakai
df = df.drop(columns=['default','day_of_week','duration','pdays','emp.var.rate','nr.employed','euribor3m'])


In [15]:
# Function filter column marital
def fill_marital(x):
    if x == 'unknown' :
        return 'married'
    else:
        return x

In [16]:
# Filter missing value column marital
df['marital'] = df['marital'].apply(fill_marital)

In [17]:
# Function filter column Job
def fill_job(x):
    if x == 'unknown' :
        return 'admin.'
    else:
        return x

In [18]:
# Filter missing value column job
df['job'] = df['job'].apply(fill_job)

In [19]:
# Function filter column Education
def fill_education(x):
    if x == 'unknown' :
        return 'university.degree'
    else:
        return x

In [20]:
# Filter missing value column education
df['edu'] = df['edu'].apply(fill_education)

In [21]:
# Melakukan Fill missing value pada kolom housing
def fill_housing(x):
    if x == 'unknown' :
        return 'yes'
    else:
        return x

In [22]:
# Filter missing value column housing
df['housing'] = df['housing'].apply(fill_housing)

In [23]:
# Melakukan Fill missing value pada kolom loan
def fill_loan(x):
    if x == 'unknown' :
        return 'no'
    else:
        return x

In [24]:
# Filter missing value column loan
df['loan'] = df['loan'].apply(fill_loan)

# **EDA**

## BUSINESS QUESTION 1 
Apakah ketika nasabah sering dihubungi jadi lebih tertarik untuk melakukan deposit atau tidak?

In [25]:
# Hasil tabular
pd.crosstab(index=df['campaign_range'], columns=df['y'], margins=True)

y                 no   yes   All
campaign_range                  
1-10            7073  1697  8770
11-20             38     1    39
21-30              1     0     1
All             7112  1698  8810

In [26]:
(pd.crosstab(index=df['campaign_range'], columns=df['y'],normalize='index', margins=True)*100).round(2)

y                   no    yes
campaign_range               
1-10             80.65  19.35
11-20            97.44   2.56
21-30           100.00   0.00
All              80.73  19.27

INSIGHT :

Kebanyakan customer sudah menentukan ingin melakukan deposit sebelum 11x dihubungi. 
Campaign yang terlalu sering atau agresif dapat membuat nasabah tidak tertarik dan bingung ( Source : https://hbr.org/2012/05/to-keep-your-customers-keep-it-simple).
sehingga setelah lebih dari 10x dihubungi dan jika belum ada tanggapan dari nasabah, tim sales dapat mencari calon nasabah lain untuk dihubungi 

Saran :

Karena peluang dihubungi 1-10 dan menjawab yes sebesar 19.35% ,maka tiap nasabah yang dihubungi 1-10 kali 
dapat meningkatkan nasabah yang ingin melakukan deposito menjadi 1.705 orang atau meningkat 0.1% 


## BUSINESS QUESTION 2 
Apakah lebih efektif melakukan campaign melalui telepon atau Cellular?


In [29]:
# Hasil Tabular
(pd.crosstab(index=df['contact'], columns=df['y'], normalize='columns')*100).round(2)

y             no    yes
contact                
cellular   90.85  93.99
telephone   9.15   6.01

INSIGHT : 

Lebih banyak orang yang mau untuk melakukan deposito saat dihubungi melalui cellular yaitu 94% dari seluruh orang yang setuju melakukan deposito. Hal ini terjadi karena sekitar 1/3 dari penduduk portugal menggunakan cellular dalam berkomunikasi pada saat itu
(http://www.websm.org/db/12/13088/Bibliography/The%20Mobile-only%20Population%20in%20Portugal%20and%20Its%20Impact%20in%20a%20Dual%20Frame%20Telephone%20Survey/?&cat=145&p1=1123&p2=82&p3=85&p4=88&p5=179&id=179&cat=145&page=5)

Saran :

Karena deposito merupakan salah satu investasi dalam jumlah yang besar, maka tim sales juga dapat mengenalkan produk deposito melalui text message terkait detail produk agar 
nasabah memiliki waktu yang cukup untuk mempelajari produk deposito
(Source : https://www.shopkick.com/partners/blog/understanding-the-types-of-consumer-buying-behavior-and-ways-to-influence-them-slp-fc)


## BUSINESS QUESTION 3
Bulan apa yang tepat untuk melakukan campaign?

In [32]:
# Hasil Tabular
pd.crosstab(index=df['month'], columns=df['y'], margins=True).sort_values('yes',ascending=False)


y        no   yes   All
month                  
All    7112  1698  8810
may    5270   524  5794
aug     506   264   770
jun     451   264   715
oct     267   180   447
nov     189   168   357
mar     156   126   282
sep     161   106   267
jul     112    66   178

In [33]:
(pd.crosstab(index=df['month'] == 'may', columns=df['y'], normalize= 'index', margins=True)*100).round(2)

y         no    yes
month              
False  61.07  38.93
True   90.96   9.04
All    80.73  19.27

In [34]:
(pd.crosstab(index=df['month'] == 'oct', columns=df['y'], normalize= 'index', margins=True)*100).round(2)

y         no    yes
month              
False  81.85  18.15
True   59.73  40.27
All    80.73  19.27

In [35]:
df_cci = pd.DataFrame({
    'Month' : ['Maret','May','Juni','Juli','Agustus','September','Oktober','November'],
    'CCI' : [-50,-46.2,-40.8,-33.6,-31.4,-29.8,-26.9,-30.1]
})
df_cci.set_index('Month').sort_values('CCI',ascending = False)

CCI
Month          
Oktober   -26.9
September -29.8
November  -30.1
Agustus   -31.4
Juli      -33.6
Juni      -40.8
May       -46.2
Maret     -50.0

INSIGHT :

Nasabah yang paling banyak untuk setuju melakukan deposit terdapat pada bulan Mei. Tetapi pada bulan Mei memiliki persentase keberhasilan 9% dari total yang dihubungi saat itu, 
hal itu terjadi karena rendah nya customer confidence index yang mempengaruhi nasabah untuk mengurangi melakukan deposit (source : https://www.investopedia.com/terms/c/cci.asp ).

Saran :

Meningkatkan jumlah nasabah yang dihubungi pada bulan yang memiliki tingkat consumer confidence index yang tinggi yaitu pada bulan oktober

## BUSINESS QUESTION 4
Apakah dengan melakukan deposit melalui digital bank dapat meningkatkan customer? 

In [60]:
# Hasil Tabular
(pd.crosstab(index=[df['housing'], df['loan']], columns=df['y'], normalize='columns', margins=True)*100).round(2)

y                no    yes    All
housing loan                     
no      no    35.26  37.28  35.65
        yes    5.71   5.95   5.75
yes     no    48.93  48.65  48.88
        yes   10.10   8.13   9.72

In [61]:
pd.crosstab(index=[df['housing'], df['loan']], columns=df['y'], margins=True)

y               no   yes   All
housing loan                  
no      no    2508   633  3141
        yes    406   101   507
yes     no    3480   826  4306
        yes    718   138   856
All           7112  1698  8810

INSIGHT :

Dari tabel di atas dapat dilihat bahwa orang yang tidak memiliki pinjaman (loan) memiliki kecenderungan untuk melakukan deposit sebesar 37,28% untuk yang tidak memiliki cicilan rumah 
dan 46,57% untuk yang memiliki cicilan rumah. Tim sales juga tidak terlalu banyak menghubungi orang yang memiliki pinjaman untuk kebutuhan pribadi

SARAN :

Tim Sales dapat menghubungi nasabah yang memiliki pinjaman dengan credit score yang baik,  hal itu dikarenakan semakin baik credit score nasabah maka semakin besar juga pendapatan nya, 
sehingga nasabah dengan credit score yang baik memiliki kecenderungan untuk melakukan deposit.
(source : https://www.americanexpress.com/en-us/credit-cards/credit-intel/credit-score-by-age-state/ )